cmd/stdiscosrv/replication.go

```diff
@@ -0,0 +1,202 @@
+// Copyright (C) 2018 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package main
+
+import (
+	"encoding/binary"
+	io "io"
+	"log"
+	"net"
+	"time"
+)
+
+type replicator interface {
+	send(key string, ps []DatabaseAddress, seen int64)
+}
+
+// a replicationSender tries to connect to the remote address and provide
+// them with a feed of replication updates.
+type replicationSender struct {
+	dst   string
+	inbox chan ReplicationRecord
+	stop  chan struct{}
+}
+
+func newReplicationSender(dst string) *replicationSender {
+	return &replicationSender{
+		dst:   dst,
+		inbox: make(chan ReplicationRecord, 1024),
+		stop:  make(chan struct{}),
+	}
+}
+
+func (s *replicationSender) Serve() {
+	// Sleep a little at startup. Peers often restart at the same time, and
+	// this avoid the service failing and entering backoff state
+	// unnecessarily, while also reducing the reconnect rate to something
+	// reasonable by default.
+	time.Sleep(2 * time.Second)
+
+	conn, err := net.Dial("tcp", s.dst)
+	if err != nil {
+		log.Println("Replication connect:", err)
+		return
+	}
+	defer conn.Close()
+
+	buf := make([]byte, 1024)
+	for {
+		select {
+		case rec := <-s.inbox:
+			// Buffer must hold record plus four bytes for size
+			size := rec.Size()
+			if len(buf) < size+4 {
+				buf = make([]byte, size+4)
+			}
+
+			// Record comes after the four bytes size
+			n, err := rec.MarshalTo(buf[4:])
+			if err != nil {
+				replicationSendsTotal.WithLabelValues("error").Inc()
+				log.Println("Replication marshal:", err)
+				continue
+			}
+			binary.BigEndian.PutUint32(buf, uint32(n))
+
+			// Send
+			conn.SetWriteDeadline(time.Now().Add(5 * time.Second))
+			if _, err := conn.Write(buf[:4+n]); err != nil {
+				replicationSendsTotal.WithLabelValues("error").Inc()
+				log.Println("Replication write:", err)
+				return
+			}
+			replicationSendsTotal.WithLabelValues("success").Inc()
+
+		case <-s.stop:
+			return
+		}
+	}
+}
+
+func (s *replicationSender) Stop() {
+	close(s.stop)
+}
+
+func (s *replicationSender) send(key string, ps []DatabaseAddress, seen int64) {
+	item := ReplicationRecord{
+		Key:       key,
+		Addresses: ps,
+	}
+	select {
+	case s.inbox <- item:
+	default:
+		replicationSendsTotal.WithLabelValues("drop").Inc()
+	}
```

Comment:
This is shit, our lookups become probabilistic

Reply:
They've always been, but in practice this never happens. Nonetheless it's the right thing to do because we don't want to block any local operations due to a stalled replication

Reply:
We should have some sort of periodic reconcile in that case?

Reply:
Nvm read the comment below.